consensus/chain_selector.go

```diff
@@ -0,0 +1,152 @@
+package consensus
+
+// This is to implement Expected Consensus protocol
+// See: https://github.com/filecoin-project/specs/blob/master/expected-consensus.md
+
+import (
+	"bytes"
+	"context"
+	"math/big"
+	"strings"
+
+	"github.com/ipfs/go-cid"
+	"github.com/ipfs/go-hamt-ipld"
+
+	"github.com/filecoin-project/go-filecoin/actor/builtin"
+	"github.com/filecoin-project/go-filecoin/state"
+	"github.com/filecoin-project/go-filecoin/types"
+)
+
+// ChainSelector weighs and compares chains according to the Storage Power
+// Consensus Protocol
+type ChainSelector struct {
+	cstore     *hamt.CborIpldStore
+	actorState SnapshotGenerator
+	genesisCid cid.Cid
+}
+
+// NewChainSelector is the constructor for chain selection module.
+func NewChainSelector(cs *hamt.CborIpldStore, actorState SnapshotGenerator, gCid cid.Cid) *ChainSelector {
+	return &ChainSelector{
+		cstore:     cs,
+		actorState: actorState,
+		genesisCid: gCid,
+	}
+}
+
+// Weight returns the expected consensus weight of this TipSet in uint64
+// encoded fixed point representation.
+func (c *ChainSelector) Weight(ctx context.Context, ts types.TipSet, pSt state.Tree) (uint64, error) {
+	ctx = log.Start(ctx, "Expected.Weight")
+	log.LogKV(ctx, "Weight", ts.String())
+	if ts.Len() == 1 && ts.At(0).Cid().Equals(c.genesisCid) {
+		return uint64(0), nil
+	}
```

Comment:
You only moved this here, but for follow-up, this should probably be a function of height rather than coupled to specific genesis CID.